clientconn.go

```diff
@@ -272,38 +278,150 @@ func DialContext(ctx context.Context, target string, opts ...DialOption) (conn *
+	// built (or right after), and this is handled in cc.updateResolverState.
+	// Also, an error from the resolver might lead to a re-resolution request
+	// from the balancer, which is handled in resolveNow() where
+	// `cc.resolverWrapper` is accessed. Hence, we need to hold the lock here.
 	cc.mu.Lock()
-	cc.resolverWrapper = rWrapper
+	cc.resolverWrapper = rw
 	cc.mu.Unlock()
 
+	// Configure idleness support with configured idle_timeout or default.
+	if cc.dopts.idleTimeout == 0 {
+		cc.idlenessMgr = newDisabledIdlenessManager()
+	} else {
+		// cc.idlenessMgr = newMutexIdlenessManager(cc, cc.dopts.idleTimeout)
+		cc.idlenessMgr = newAtomicIdlenessManager(cc, cc.dopts.idleTimeout)
+	}
+
+	// Return early for non-blocking dials.
+	if !cc.dopts.block {
+		return cc, nil
+	}
+
 	// A blocking dial blocks until the clientConn is ready.
-	if cc.dopts.block {
-		for {
+	for {
+		s := cc.GetState()
+		if s == connectivity.Idle {
 			cc.Connect()
-			s := cc.GetState()
-			if s == connectivity.Ready {
-				break
-			} else if cc.dopts.copts.FailOnNonTempDialError && s == connectivity.TransientFailure {
-				if err = cc.connectionError(); err != nil {
-					terr, ok := err.(interface {
-						Temporary() bool
-					})
-					if ok && !terr.Temporary() {
-						return nil, err
-					}
-				}
-			}
-			if !cc.WaitForStateChange(ctx, s) {
-				// ctx got timeout or canceled.
-				if err = cc.connectionError(); err != nil && cc.dopts.returnLastError {
+		}
+		if s == connectivity.Ready {
+			return cc, nil
+		} else if cc.dopts.copts.FailOnNonTempDialError && s == connectivity.TransientFailure {
+			if err = cc.connectionError(); err != nil {
+				terr, ok := err.(interface {
+					Temporary() bool
+				})
+				if ok && !terr.Temporary() {
 					return nil, err
 				}
-				return nil, ctx.Err()
 			}
 		}
+		if !cc.WaitForStateChange(ctx, s) {
+			// ctx got timeout or canceled.
+			if err = cc.connectionError(); err != nil && cc.dopts.returnLastError {
+				return nil, err
+			}
+			return nil, ctx.Err()
+		}
+	}
+}
+
+// addTraceEvent is a helper method to add a trace event on the channel. If the
+// channel is a nested one, the same event is also added on the parent channel.
+func (cc *ClientConn) addTraceEvent(msg string) {
+	ted := &channelz.TraceEventDesc{
+		Desc:     fmt.Sprintf("Channel %s", msg),
+		Severity: channelz.CtInfo,
+	}
+	if cc.dopts.channelzParentID != nil {
+		ted.Parent = &channelz.TraceEventDesc{
+			Desc:     fmt.Sprintf("Nested channel(id:%d) %s", cc.channelzID.Int(), msg),
+			Severity: channelz.CtInfo,
+		}
+	}
+	channelz.AddTraceEvent(logger, cc.channelzID, 0, ted)
+}
+
+// exitIdleMode moves the channel out of idle mode by recreating the name
+// resolver and load balancer.
+func (cc *ClientConn) exitIdleMode() error {
+	cc.mu.Lock()
+	if cc.conns == nil {
+		cc.mu.Unlock()
+		return errConnClosing
+	}
+	if cc.idlenessState != ccIdlenessStateIdle {
+		// TODO: Switch this to warning once idleness implementation is stable.
```

Comment:
Either:

1. We are sure this condition is impossible, in which case it should stay as an error
2. This condition may be possible, in which case we should not log at all

I don't think "warning" is appropriate here.

Reply:
In my testing so far, I have not seen this error condition being hit. So, I removed the TODO. If we hit this in the future, having the error log will fail our tests and we can investigate further.